# Feature Engineering

This notebook prepares the cleaned PCOS dataset for machine learning by selecting relevant features, handling categorical variables, splitting the data into training and testing sets, and scaling numerical features where appropriate.

In [1]:
#Importing libraries
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

In [2]:
#Dataset loading
df = pd.read_csv("../data/processed/pcos_cleaned.csv")

df.head()

,Sl. No,Patient File No.,PCOS (Y/N),Age (yrs),Weight (Kg),Height(Cm),BMI,Blood Group,Pulse rate(bpm),RR (breaths/min),...,Pimples(Y/N),Fast food (Y/N),Reg.Exercise(Y/N),BP _Systolic (mmHg),BP _Diastolic (mmHg),Follicle No. (L),Follicle No. (R),Avg. F size (L) (mm),Avg. F size (R) (mm),Endometrium (mm)
0,1,1,0,28,44.6,152.0,19.300000,15,78,22,...,0,1.0,0,110,80,3,3,18.0,18.0,8.5
1,2,2,0,36,65.0,161.5,24.921163,15,74,20,...,0,0.0,0,120,70,3,5,15.0,14.0,3.7
2,3,3,1,33,68.8,165.0,25.270891,11,72,18,...,1,1.0,0,120,80,13,15,18.0,20.0,10.0
3,4,4,0,37,65.0,148.0,29.674945,13,72,20,...,0,0.0,0,120,70,2,2,15.0,14.0,7.5
4,5,5,0,25,52.0,161.0,20.060954,11,72,18,...,0,0.0,0,120,80,3,4,16.0,14.0,7.0


In [3]:
#Dataset Shape
df.shape

(541, 44)

The cleaned dataset is loaded successfully and will be used for feature engineering prior to predictive modeling.

In [4]:
#Separating Target Variable
y = df["PCOS (Y/N)"]
X = df.drop(columns=["PCOS (Y/N)"])

In [5]:
#Removing Unnecessary Columns
X = X.drop(columns=[
    "Sl. No",
    "Patient File No."
])

In [6]:
X.head()

,Age (yrs),Weight (Kg),Height(Cm),BMI,Blood Group,Pulse rate(bpm),RR (breaths/min),Hb(g/dl),Cycle(R/I),Cycle length(days),...,Pimples(Y/N),Fast food (Y/N),Reg.Exercise(Y/N),BP _Systolic (mmHg),BP _Diastolic (mmHg),Follicle No. (L),Follicle No. (R),Avg. F size (L) (mm),Avg. F size (R) (mm),Endometrium (mm)
0,28,44.6,152.0,19.300000,15,78,22,10.48,2,5,...,0,1.0,0,110,80,3,3,18.0,18.0,8.5
1,36,65.0,161.5,24.921163,15,74,20,11.70,2,5,...,0,0.0,0,120,70,3,5,15.0,14.0,3.7
2,33,68.8,165.0,25.270891,11,72,18,11.80,2,5,...,1,1.0,0,120,80,13,15,18.0,20.0,10.0
3,37,65.0,148.0,29.674945,13,72,20,12.00,2,5,...,0,0.0,0,120,70,2,2,15.0,14.0,7.5
4,25,52.0,161.0,20.060954,11,72,18,10.00,2,5,...,0,0.0,0,120,80,3,4,16.0,14.0,7.0


In [7]:
X.shape

(541, 41)

Patient identifiers were removed because they do not contribute predictive information. The remaining variables consist of demographic, clinical, lifestyle, hormonal, and ultrasound measurements that may be useful for predicting PCOS.

In [8]:
#Encoding requirment
X.dtypes

Age (yrs)                  int64
Weight (Kg)              float64
Height(Cm)               float64
BMI                      float64
Blood Group                int64
Pulse rate(bpm)            int64
RR (breaths/min)           int64
Hb(g/dl)                 float64
Cycle(R/I)                 int64
Cycle length(days)         int64
Marraige Status (Yrs)    float64
Pregnant(Y/N)              int64
No. of aborptions          int64
I beta-HCG(mIU/mL)       float64
II beta-HCG(mIU/mL)      float64
FSH(mIU/mL)              float64
LH(mIU/mL)               float64
FSH/LH                   float64
Hip(inch)                  int64
Waist(inch)                int64
Waist:Hip Ratio          float64
TSH (mIU/L)              float64
AMH(ng/mL)               float64
PRL(ng/mL)               float64
Vit D3 (ng/mL)           float64
PRG(ng/mL)               float64
RBS(mg/dl)               float64
Weight gain(Y/N)           int64
hair growth(Y/N)           int64
Skin darkening (Y/N)       int64
Hair loss(

All categorical variables were already represented numerically in the cleaned dataset. Therefore, no additional encoding was required.

# Train-Test Split

In [9]:
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

In [10]:
print(X_train.shape)
print(X_test.shape)

(432, 41)
(109, 41)


The dataset was divided into training and testing sets using an **80:20 split** while preserving the original class distribution through **stratified sampling**.

- **Training set:** 432 patient records (80%)
- **Testing set:** 109 patient records (20%)

The training dataset will be used to build and optimize machine learning models, while the testing dataset will be reserved for evaluating model performance on unseen data. This approach helps assess how well the model generalizes to new patient data and reduces the risk of overfitting.

In [15]:
#Verifying Class Distribution
distribution = pd.DataFrame({
    "Training (%)": (y_train.value_counts(normalize=True) * 100).round(2),
    "Testing (%)": (y_test.value_counts(normalize=True) * 100).round(2)
})

distribution.index = ["No PCOS", "PCOS"]

distribution

,Training (%),Testing (%)
No PCOS,67.36,66.97
PCOS,32.64,33.03


 ### Class Distribution After Train-Test Split

The class proportions in both the training and testing datasets closely match the original dataset, confirming that stratified sampling successfully preserved the distribution of the target variable.

- Training set: 67.36% No PCOS and 32.64% PCOS
- Testing set: 66.97% No PCOS and 33.03% PCOS

Maintaining similar class distributions across both datasets helps reduce sampling bias and ensures that the machine learning models are trained and evaluated on representative data.

## Feature Scaling

Many machine learning algorithms are sensitive to differences in the scale of input features. Clinical measurements such as BMI, hormone levels, blood glucose, and follicle measurements are recorded using different units and ranges. Without scaling, variables with larger numerical values may disproportionately influence model training.

To address this, **StandardScaler** was applied to the **continuous numerical features**, transforming each feature to have a mean of approximately 0 and a standard deviation of 1. Binary and categorical variables were left unchanged because they already represent discrete categories and do not require scaling.

The scaler was fitted only on the training dataset and subsequently applied to the testing dataset to prevent data leakage.

In [20]:
# Defining continuous feature
continuous_features = [
    'Age (yrs)',
    'Weight (Kg)',
    'Height(Cm)',
    'BMI',
    'Pulse rate(bpm)',
    'RR (breaths/min)',
    'Hb(g/dl)',
    'Cycle length(days)',
    'Marraige Status (Yrs)',
    'No. of aborptions',
    'I beta-HCG(mIU/mL)',
    'II beta-HCG(mIU/mL)',
    'FSH(mIU/mL)',
    'LH(mIU/mL)',
    'FSH/LH',
    'Hip(inch)',
    'Waist(inch)',
    'Waist:Hip Ratio',
    'TSH (mIU/L)',
    'AMH(ng/mL)',
    'PRL(ng/mL)',
    'Vit D3 (ng/mL)',
    'PRG(ng/mL)',
    'RBS(mg/dl)',
    'BP _Systolic (mmHg)',
    'BP _Diastolic (mmHg)',
    'Follicle No. (L)',
    'Follicle No. (R)',
    'Avg. F size (L) (mm)',
    'Avg. F size (R) (mm)',
    'Endometrium (mm)'
]

In [21]:
X_train_scaled = X_train.copy()
X_test_scaled = X_test.copy()

In [23]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

X_train_scaled[continuous_features] = scaler.fit_transform(
    X_train[continuous_features]
)

X_test_scaled[continuous_features] = scaler.transform(
    X_test[continuous_features]
)

In [24]:
# Converting back to DataFrame
X_train_scaled = pd.DataFrame(
    X_train_scaled,
    columns=X_train.columns
)

X_test_scaled = pd.DataFrame(
    X_test_scaled,
    columns=X_test.columns
)

In [27]:
X_train_scaled[continuous_features].describe().round(2)

,Age (yrs),Weight (Kg),Height(Cm),BMI,Pulse rate(bpm),RR (breaths/min),Hb(g/dl),Cycle length(days),Marraige Status (Yrs),No. of aborptions,...,Vit D3 (ng/mL),PRG(ng/mL),RBS(mg/dl),BP _Systolic (mmHg),BP _Diastolic (mmHg),Follicle No. (L),Follicle No. (R),Avg. F size (L) (mm),Avg. F size (R) (mm),Endometrium (mm)
count,432.00,432.00,432.00,432.00,432.00,432.00,432.00,432.00,432.00,432.00,...,432.00,432.00,432.00,432.00,432.00,432.00,432.00,432.00,432.00,432.00
mean,-0.00,-0.00,0.00,0.00,-0.00,0.00,0.00,0.00,0.00,-0.00,...,0.00,0.00,0.00,-0.00,0.00,-0.00,-0.00,-0.00,0.00,0.00
std,1.00,1.00,1.00,1.00,1.00,1.00,1.00,1.00,1.00,1.00,...,1.00,1.00,1.00,1.00,1.00,1.00,1.00,1.00,1.00,1.00
min,-2.10,-2.55,-3.19,-2.89,-12.69,-1.91,-2.49,-3.25,-1.59,-0.42,...,-0.13,-0.33,-2.56,-13.25,-11.79,-1.44,-1.49,-4.05,-4.60,-3.83
25%,-0.81,-0.66,-0.72,-0.65,-0.25,-0.74,-0.77,-0.63,-0.77,-0.42,...,-0.09,-0.17,-0.48,-0.60,-1.16,-0.73,-0.81,-0.53,-0.72,-0.67
50%,-0.08,-0.04,-0.06,-0.04,-0.25,-0.74,-0.20,0.03,-0.15,-0.42,...,-0.08,-0.12,0.03,-0.60,0.55,-0.26,-0.12,0.01,0.17,0.00
75%,0.66,0.50,0.60,0.59,0.17,0.44,0.63,0.03,0.47,-0.42,...,-0.05,-0.02,0.49,0.69,0.55,0.70,0.80,0.82,0.77,0.60
max,3.06,4.35,3.89,3.58,1.86,5.12,4.16,4.62,4.59,7.15,...,15.41,19.57,8.14,3.27,3.98,3.79,3.09,2.45,2.56,4.28


In [28]:
X_train_scaled.head()

,Age (yrs),Weight (Kg),Height(Cm),BMI,Blood Group,Pulse rate(bpm),RR (breaths/min),Hb(g/dl),Cycle(R/I),Cycle length(days),...,Pimples(Y/N),Fast food (Y/N),Reg.Exercise(Y/N),BP _Systolic (mmHg),BP _Diastolic (mmHg),Follicle No. (L),Follicle No. (R),Avg. F size (L) (mm),Avg. F size (R) (mm),Endometrium (mm)
268,-1.182193,-0.529909,-0.721223,-0.267325,11,-0.249844,-0.736436,1.408343,4,-1.281682,...,0,0.0,0,-0.604025,0.552406,-0.017074,-0.347227,0.551537,1.069129,-0.671959
6,0.475948,0.410898,-0.062019,0.500192,11,-0.249844,-0.736436,-0.312517,2,0.028819,...,0,0.0,0,0.686474,0.552406,-0.017074,-0.117866,0.010151,0.173830,-0.762051
254,-0.629479,-0.664310,-0.062019,-0.704203,13,-0.249844,-1.908223,0.031655,2,0.028819,...,1,0.0,0,0.686474,-1.161957,-0.968812,-0.805949,0.010151,0.770696,-0.221498
440,-0.997955,-0.305907,0.597186,-0.964947,11,-0.671455,0.435352,-0.083069,2,0.684069,...,0,1.0,1,-0.604025,-1.161957,0.696730,-0.117866,1.092922,-0.423036,-0.221498
415,-0.076766,0.052495,0.597186,-0.183342,15,1.436602,0.435352,-0.427241,4,-1.936933,...,1,1.0,0,0.686474,0.552406,0.458795,1.258300,-0.801927,0.173830,-0.897190


### Observation

The summary statistics confirm that the continuous numerical variables were successfully standardized, with means close to **0** and standard deviations close to **1**. Binary and categorical variables, including clinical indicators such as pimples, fast food consumption, regular exercise, and blood group, were intentionally left unchanged to preserve their original categorical representation.

This preprocessing strategy ensures that continuous features are placed on a comparable scale while maintaining the interpretability of binary variables. The resulting dataset is well prepared for training and evaluating machine learning models in the next stage of the project.

In [13]:
# Saving processed data
X_train_scaled.to_csv("../data/processed/X_train.csv", index=False)

X_test_scaled.to_csv("../data/processed/X_test.csv", index=False)

y_train.to_csv("../data/processed/y_train.csv", index=False)

y_test.to_csv("../data/processed/y_test.csv", index=False)

## Summary

In this notebook, the cleaned PCOS dataset was prepared for machine learning.

- The target variable was separated from the predictor variables.
- Patient identifier columns were removed, as they do not contribute predictive information.
- The remaining demographic, clinical, hormonal, lifestyle, and ultrasound features were retained for model development.
- The dataset was divided into training and testing sets using an **80:20 stratified split** to preserve the original class distribution.
- Continuous numerical features were standardized using **StandardScaler**, while binary and categorical variables were left unchanged.
- The processed training and testing datasets were saved for use in the machine learning pipeline.

The dataset is now fully prepared for building, training, and evaluating predictive models in the next notebook.